src/Npgsql/ConnectorPool.Multiplexing.cs

```diff
@@ -0,0 +1,459 @@
+        /// flushing to the network.
+        /// </summary>
+        readonly int _writeCoalescingBufferThresholdBytes;
+
+        internal long NumCommandsSent;
+        internal long NumBatches;
+        long _ticksWritten;
+        long _waitsForFurtherCommands;
+        // long _bytesFlushed;
+
+        /// <summary>
+        /// Called exactly once per multiplexing pool, when the first connection is opened, with two goals:
+        /// 1. Load types and bind the pool-wide type mapper (necessary for binding parameters)
+        /// 2. Cause any connection exceptions (e.g. bad username) to be thrown from NpgsqlConnection.Open
+        /// </summary>
+        internal async Task BootstrapMultiplexing(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken = default)
+        {
+            Debug.Assert(_multiplexing);
+
+            var connector =
+                await conn.StartBindingScope(ConnectorBindingScope.Connection, timeout, async, cancellationToken);
+            using var _ = Defer(() => conn.EndBindingScope(ConnectorBindingScope.Connection));
+
+            // Somewhat hacky. Extract the connector's type mapper as our pool-wide mapper,
+            // and have the connector rebind to ensure it has a different instance.
+            // The latter isn't strictly necessary (type mappers should always be usable
+            // concurrently) but just in case.
+            MultiplexingTypeMapper = connector.TypeMapper;
+            connector.RebindTypeMapper();
+
+            IsBootstrapped = true;
+        }
+
+        async Task MultiplexingWriteLoopWrapper()
+        {
+            try
+            {
+                await MultiplexingWriteLoop();
+            }
+            catch (Exception e)
+            {
+                Log.Error("Exception in multiplexing write loop, this is an Npgsql bug, please file an issue.", e);
+            }
+        }
+
+        async Task MultiplexingWriteLoop()
+        {
+            // This method is async, but only ever yields when there are no pending commands in the command channel.
+            // No I/O should ever be performed asynchronously, as that would block further writing for the entire
+            // application; whenever I/O cannot complete immediately, we chain a callback with ContinueWith and move
+            // on to the next connector.
+            Debug.Assert(_multiplexCommandReader != null);
+
+            var timeout = _writeCoalescingDelayTicks / 2;
+            var timeoutTokenSource = new CancellationTokenSource();
+            var timeoutToken = timeout == 0 ? CancellationToken.None : timeoutTokenSource.Token;
+
+            // TODO: Writing I/O here is currently async-only. Experiment with both sync and async (based on user
+            // preference, ExecuteReader vs. ExecuteReaderAsync).
+            while (true)
+            {
+                var stats = new MultiplexingStats { Stopwatch = new Stopwatch() };
```

Comment:
Nice, did not know that existed. I don't think there would be a whole lot of value in this specific case, though. The way the loop is designed, is that as long as write I/O happens synchronously - which should be the case except if the database is totally backed up - we reset and reuse the same Stopwatch instance; we only allocate a new Stopwatch if a write operation completes asynchronously, in which case I'm reasoning that the extra allocation is negligible. But if you think otherwise let me know.

In general, if you look at this loop, you'll see that the synchronously-completing paths are quite optimized, whereas the asynchronously-completing ones are much less so.

Reply:
Agreed, it's just a little helper that's easily worth the small size increase if you use it all over the codebase.